### PR TITLE
Add quality declaration for foonathan_memory

### PR DESCRIPTION
**Pull request by brawner**
This provides a Quality Declaration for the foonathan_memory third-party package as part of the [REP-2004](https://discourse.ros.org/t/rep-2004-package-quality-categories/11988) effort. It currently declares this package as Quality Level 3, which is based on the information I could find. However, I'm perfectly happy to make changes if people have other input to provide.

### CODE REVIEW

**Comment by brawner** [comment]
Related issue for tracking progress to QL 1 https://github.com/eProsima/foonathan_memory_vendor/issues/23

**Comment by brawner** [comment]
This PR updated the QD to follow the current organizational pattern we're following for QDs.

**Review by nuclearsandwich — DISMISSED** [review]
I've reviewed this quality declaration as a maintainer of the vendor package. It appears correct with respect to the information upstream. I haven't checked that it conforms to the REP-2004 standard although it looks like it does. I wouldn't mind if another contributor working on Quality Declarations reviews here.

File: QUALITY_DECLARATION.md
Comment (Blast545) [inline QUALITY_DECLARATION.md]: Even though this is true, as of #24 the vendor package imports a specific version of foonathan_memory, so I think an argument can be made to say that fixing the version shouldn't introduce problems related to API/ABI stability (correct me if I'm wrong). I made a similar one for `libyaml` you can check it [here](https://github.com/eProsima/foonathan_memory_vendor/pull/24)

File: QUALITY_DECLARATION.md
Comment (Blast545) [inline QUALITY_DECLARATION.md]: Mind adding table with the summary of each of the topics? It will help external readers with the sections of the Quality Levels partially addressed. (For example, the API testing which is partially addressed)

File: QUALITY_DECLARATION.md
Comment (Blast545) [inline QUALITY_DECLARATION.md]: Would you mind adding a summary section with your thoughts about what would be needed for this package to be considered Quality Level 1 in terms of ROS2 usage? You can check the one I wrote for `libyaml` [here](https://github.com/ros2/libyaml_vendor/pull/14) as an example for this. 

This section is added for external dependencies as many won't be matching 100% the [REP-2004](https://github.com/ros-infrastructure/rep/pull/218) requirements and still might be considered Quality Level 1 on a case by case basis.

**Review by Blast545 — COMMENTED** [review]
Overall LGTM! let me know your thoughts on the comments I left you

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: Added a sentence clarifying this information.

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: A summary of its current progress can be found in the related issue (https://github.com/eProsima/foonathan_memory_vendor/issues/23).

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: Added summary section (8) at end of the QD.

File: QUALITY_DECLARATION.md
Comment (Blast545) [inline QUALITY_DECLARATION.md]:
```suggestion
Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 4 in REP-2004](https://www.ros.org/reps/rep-2004.html).
```

File: QUALITY_DECLARATION.md
Comment (Blast545) [inline QUALITY_DECLARATION.md]: I think this is a typo
```suggestion
If a newer version is desired for a previously released ROS 2 distribution, then it will need to be verified that the API does not change between versions.
```

**Review by Blast545 — DISMISSED** [review]
Looks good to me! I liked the steps marked at the end to be followed.

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: Fixed

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: Fixed.

**Review by Blast545 — DISMISSED** [review]

File: QUALITY_DECLARATION.md
Comment (nuclearsandwich) [inline QUALITY_DECLARATION.md]: Where does the "beta" declaration come from upstream? I guess I'm asking whether there's a source to cite for this or if it's based on implicit information and what information if so?

That the version number is below 1.0 communicates some meaning but nothing specific since there is no declared version scheme.

**Review by nuclearsandwich — DISMISSED** [review]
I have one last question that I forgot to register last time through. I would like to get an answer but I don't believe it blocks the PR. To avoid losing the question in a merged PR I'll wait to merge until the answer or end-of-day today whichever comes first.

File: QUALITY_DECLARATION.md
Comment (ahcorde) [inline QUALITY_DECLARATION.md]:
```suggestion
### Contributor Origin [2.ii]

This package does not have a confirmation of contributor origin policy.
```

File: QUALITY_DECLARATION.md
Comment (ahcorde) [inline QUALITY_DECLARATION.md]: Ubuntu 16.04 ?  we are targeting Focal, (20.04) right?

**Review by ahcorde — CHANGES_REQUESTED** [review]
LGTM. But I have a question regarting the Ubuntu version

File: QUALITY_DECLARATION.md
Comment (nuclearsandwich) [reply QUALITY_DECLARATION.md]: My understanding is that this is documenting foonathan_memory's CI upstream. Which is using 16.04.

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: beta is probably not the right word, since this package is pretty much in maintenance mode. I adjusted the language to reflect that API changes may occur during small version updates. Also added a cite to the changelog.

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: Fixed.

File: QUALITY_DECLARATION.md
Comment (brawner) [reply QUALITY_DECLARATION.md]: Err, I see why its confusing since I just named this QUALITY_DECLARATION.md. I renamed the file to reflect that it's for foonathan_memory itself, and not the _vendor package.

**Review by nuclearsandwich — APPROVED** [review]

**Comment by nuclearsandwich** [comment]
I believe @ahcorde's only blocking question in https://github.com/eProsima/foonathan_memory_vendor/pull/22#discussion_r417344364 has been addressed satisfactorily. I'm going to go ahead and merge this and in doing so take co-ownership of the resolution if I turn out to be incorrect.

**Pull request merged**
